src/main.py

```diff
@@ -237,8 +237,8 @@ def parse_transaction(b64_tx, success_callback=None, error_callback=None):
         for el in scope:
             for der in el.bip32_derivations:
```

Comment:
weird. On stm32 i cant get into this loop because el.bip32_derivations is empty. Works in simulator though.
I copied the lines from you:
 ```
    # blue wallet trick - if the fingerprint is 0 we use our fingerprint
    for scope in [tx.inputs, tx.outputs]:
        for el in scope:
            for der in el.bip32_derivations:
                if el.bip32_derivations[der].fingerprint == b'\x00\x00\x00\x00':
                    el.bip32_derivations[der].fingerprint = keystore.fingerprint
```
and i wanted to ask you if you tested it on hw, but it seems this code doesnt do anything, its dead.

Reply:
It's not dead. It works only if fingerprint in bip32 derivation of PSBT is zero - Bitcoin Core will never produce such transaction, but Blue Wallet doesn't know our fingerprint so they just fill zeroes there. If you try with Blue Wallet or manually set fingerprint to \x00\x00\x00\x00 you will see that this code is actually replacing zero fingerprint with the fingerprint of the device.

Reply:
And yeah, I tested that tx signing with blue wallet works, so it should enter this loop.
bip39 derivation is empty probably because we delete all unnecessary fields after signing.

Reply:
Here we delete bip39 derivation. 
https://github.com/cryptoadvance/specter-diy/pull/59/files#diff-0c81b9c94c68ea34c497c1fa52daee56R194
So you need to do index update before this happens

Reply:
what am i missing then. Who/what is using variable *el*?

Reply:
It fills the fingerprint in the original transaction. 
`el` is an instalnce of PSBTInput/Output in the list, `bip32_derivation` is a dictionary.
So by when class instance is copied to the variable `el`,  `bip32_derivation` is not a copy of origianl dict, but a reference. 
So by changing `el.bip32_derivations[der].fingerprint` we are actually changing corresponding part in the original transaction: `tx.inputs[i].bip32_derivations[der].fingerprint`.

I agree that this piece of code is very confusing.
It might be more understandable if we change input and output directly using index with `enumerate` and maybe use two identical loops through inputs and outputs.

Reply:
Amazing.

> so it should enter this loop.

For me it doesnt enter that loop. This code asserts only in simulator, but not on hardware. I scan exactly the same transaction on both platforms and both platforms contain the same userdata folder. Anyways, ill debug tomorrow and try to resolve this...

```
    # blue wallet trick - if the fingerprint is 0 we use our fingerprint
    for scope in [tx.inputs, tx.outputs]:
        for el in scope:
            print ("+++++++++++++")
            for der in el.bip32_derivations:
                assert(0)
                if el.bip32_derivations[der].fingerprint == b'\x00\x00\x00\x00':
                    el.bip32_derivations[der].fingerprint = keystore.fingerprint
```

Reply:
I figured out the problem: 

![scr](https://user-images.githubusercontent.com/25270775/82199564-23ebea00-98fe-11ea-95f4-28bcb7f04366.png)

I thought the payloads in QRcode and in text were supposed to be the same.

COPIED:

`cHNidP8BAIkCAAAAAdGUzLtZzKFFaKwlJ/6i9SoClwIl9D39o4jJ0GGREHypAAAAAAD+////AlQKDY8AAAAAIgAgZECaMwzPqUrlqKQV1f13LR4tJB4wRm6k8fJJMw5jBPAA4fUFAAAAACIAIDDbsdJT8/jJ9Hc9h9gkaQxESiRM4H12HT13roHnvWITAAAAAE8BBDWHzwQixZjegAAAAiHVE6AJMxGZaJzTEtGe8d52W3DDHq1m2GWcxebVQeYbApwt0keBnXaiR+HPD9AGmGu44S7DXpwG6G1irgMkJw++FJ0dxgSAAAAwgAAAAYAAAACAAAACTwEENYfPBIwihT6AAAACEttNnBEfJ/EDUr86z1GCgYFvFzSgm0bbYg93s72I4OsCxDvJhefZEcPqBtFqvvQr4IYV7brPRy+hU8nzX35QULwU+HIPjoAAADCAAAABgAAAAIAAAAIAAQErAPkClQAAAAAiACBnx0f/wIzvwdFA7bRBfNE50U/dhy7RY7umbJIBHdz8owEFR1EhA3syvd5QtweU1o2w6vZmgcUT1/+Qi2s771bzIVmz2+kuIQPh4V5YF2smwV+3l+UolCro0t/ZgsHgS5mWwikzsp5k+FKuIgYDezK93lC3B5TWjbDq9maBxRPX/5CLazvvVvMhWbPb6S4cnR3GBDAAAIABAACAAAAAgAIAAIAAAAAAAAAAACIGA+HhXlgXaybBX7eX5SiUKujS39mCweBLmZbCKTOynmT4HPhyD44wAACAAQAAgAAAAIACAACAAAAAAAAAAAAAAQFHUSECAQbB7bMRODRTRLfxIZ/eFqaOQYF1/rV51OGifKmEj7YhA9qgAEktNr5tUR0CKTxDkKXhkEuFwr+f4PbjaOUj3JywUq4iAgIBBsHtsxE4NFNEt/Ehn94Wpo5BgXX+tXnU4aJ8qYSPthz4cg+OMAAAgAEAAIAAAACAAgAAgAEAAAAAAAAAIgID2qAASS02vm1RHQIpPEOQpeGQS4XCv5/g9uNo5SPcnLAcnR3GBDAAAIABAACAAAAAgAIAAIABAAAAAAAAAAABAUdRIQJAHnTn9EHrzwQSrJL4Kzw0KY+vNYRcXUZdL04stGLAPyEDEF1dAqlVsS5FHisIs+YbaR0BWwNhWjQRaQl7fEIlMhxSriICAkAedOf0QevPBBKskvgrPDQpj681hFxdRl0vTiy0YsA/HPhyD44wAACAAQAAgAAAAIACAACAAAAAAAwAAAAiAgMQXV0CqVWxLkUeKwiz5htpHQFbA2FaNBFpCXt8QiUyHBydHcYEMAAAgAEAAIAAAACAAgAAgAAAAAAMAAAAAA==`


SCANNED:

`cHNidP8BAIkCAAAAAdGUzLtZzKFFaKwlJ/6i9SoClwIl9D39o4jJ0GGREHypAAAAAAD+////AlQKDY8AAAAAIgAgZECaMwzPqUrlqKQV1f13LR4tJB4wRm6k8fJJMw5jBPAA4fUFAAAAACIAIDDbsdJT8/jJ9Hc9h9gkaQxESiRM4H12HT13roHnvWITAAAAAAABASsA+QKVAAAAACIAIGfHR//AjO/B0UDttEF80TnRT92HLtFju6ZskgEd3PyjB/zKASOiEV8IAAAAAAAAAAAAB/zKASOiEV8IAQAAAAAAAAAAB/zKASOiEV8IAAAAAAwAAAAA`

Reply:
There is a special custom field in compressed psbt that contains derivation index.
It is stored in "unknown" dict, the key value is b"\xfc\xca\x01"+wallet.fingerprint, derivation path is two 4-byte indexes - change and address index.

Here is an example how I process it:
https://github.com/cryptoadvance/specter-diy/blob/master/src/keystore.py#L330